File: src/lib/java/work/lclpnet/ap2/impl/game/kit/KitHandle.java
package work.lclpnet.ap2.impl.game.kit;

import net.minecraft.core.RegistryAccess;
import net.minecraft.core.component.DataComponents;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.Identifier;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.component.ItemLore;
import net.minecraft.world.item.component.TooltipDisplay;
import work.lclpnet.ap2.impl.util.IconMaker;
import work.lclpnet.kibu.hook.HookRegistrar;
import work.lclpnet.kibu.inv.item.ItemStackUtil;
import work.lclpnet.kibu.scheduler.api.TaskScheduler;
import work.lclpnet.kibu.translate.Translations;
import work.lclpnet.kibu.translate.text.RootText;
import work.lclpnet.kibu.translate.text.TranslatedText;

import java.util.ArrayList;
import java.util.List;

import static net.minecraft.ChatFormatting.AQUA;
import static net.minecraft.ChatFormatting.GREEN;

public interface KitHandle {

    Identifier gameId();

    HookRegistrar hooks();

    TaskScheduler scheduler();

    Translations translations();

    RegistryAccess registries();

    KitReadView readView();

    default boolean hasKitEquipped(ServerPlayer player, Kit kit) {
        return readView().hasKitEquipped(player, kit);
    }

    default ItemStack createKitIcon(Kit kit, ServerPlayer player) {
        ItemStack stack = kit.createItemStack(registries());

        decorateItemStack(stack, kit, player, true);

        return stack;
    }

    default ItemStack createItemStack(Kit kit, ServerPlayer player) {
        ItemStack stack = kit.createItemStack(registries());

        decorateItemStack(stack, kit, player, false);

        return stack;
    }

    default void decorateItemStack(ItemStack stack, Kit kit, ServerPlayer player, boolean forIcon) {
        Identifier gameId = gameId();
        Translations translations = translations();

        stack.set(DataComponents.CUSTOM_NAME, kitName(kit).translateFor(player).formatted(AQUA)
                .styled(style -> style.withItalic(false)));

        stack.set(DataComponents.TOOLTIP_DISPLAY, TooltipDisplay.DEFAULT
                .withHidden(DataComponents.ATTRIBUTE_MODIFIERS, true)
                .withHidden(DataComponents.UNBREAKABLE, true)
                .withHidden(DataComponents.ENCHANTMENTS, true)
                .withHidden(DataComponents.DAMAGE, true));

        String descriptionPath = forIcon ? "description" : "hint";
        String descriptionKey = "game.%s.%s.kit.%s.%s"
                .formatted(gameId.getNamespace(), gameId.getPath(), kit.id(), descriptionPath);

        if (!translations.getTranslator().hasTranslation(translations.getLanguage(player), descriptionKey)) return;

        RootText description = translations.translateText(player, descriptionKey).formatted(GREEN);

        List<Component> currentLore = stack.getOrDefault(DataComponents.LORE, ItemLore.EMPTY).styledLines();
        List<Component> loreToAdd = IconMaker.wrapText(description, 32);
        List<Component> newLore = new ArrayList<>(currentLore);

        if (!currentLore.isEmpty()) {
            // newline between the lore
            newLore.add(Component.nullToEmpty(""));
        }

        newLore.addAll(loreToAdd);

        ItemStackUtil.setLore(stack, newLore);
    }

    default TranslatedText kitName(Kit kit) {
        Identifier gameId = gameId();

        return translations().translateText("game.%s.%s.kit.%s".formatted(gameId.getNamespace(), gameId.getPath(), kit.id()));
    }
}

File: src/lib/src/main/kotlin/work/lclpnet/ap2/game/kit/KitHandle.kt
package work.lclpnet.ap2.game.kit

import net.minecraft.ChatFormatting
import net.minecraft.core.RegistryAccess
import net.minecraft.core.component.DataComponents
import net.minecraft.network.chat.Component
import net.minecraft.resources.Identifier
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.component.ItemLore
import net.minecraft.world.item.component.TooltipDisplay
import work.lclpnet.ap2.impl.util.IconMaker
import work.lclpnet.kibu.hook.HookRegistrar
import work.lclpnet.kibu.inv.item.ItemStackUtil
import work.lclpnet.kibu.scheduler.api.TaskScheduler
import work.lclpnet.kibu.translate.Translations
import work.lclpnet.kibu.translate.text.TranslatedText

interface KitHandle {
    val gameId: Identifier
    val hooks: HookRegistrar
    val scheduler: TaskScheduler
    val translations: Translations
    val registries: RegistryAccess
    val readView: KitReadView

    fun hasKitEquipped(player: ServerPlayer, kit: Kit): Boolean {
        return readView.hasKitEquipped(player, kit)
    }

    fun createKitIcon(kit: Kit, player: ServerPlayer): ItemStack {
        val stack = kit.createItemStack(registries)

        decorateItemStack(stack, kit, player, true)

        return stack
    }

    fun createItemStack(kit: Kit, player: ServerPlayer): ItemStack {
        val stack = kit.createItemStack(registries)

        decorateItemStack(stack, kit, player, false)

        return stack
    }

    fun decorateItemStack(stack: ItemStack, kit: Kit, player: ServerPlayer, forIcon: Boolean) {
        stack.set(
            DataComponents.CUSTOM_NAME, kitName(kit).translateFor(player).formatted(ChatFormatting.AQUA)
                .styled { style -> style.withItalic(false) }
        )

        stack.set(
            DataComponents.TOOLTIP_DISPLAY, TooltipDisplay.DEFAULT
                .withHidden(DataComponents.ATTRIBUTE_MODIFIERS, true)
                .withHidden(DataComponents.UNBREAKABLE, true)
                .withHidden(DataComponents.ENCHANTMENTS, true)
                .withHidden(DataComponents.DAMAGE, true)
        )

        val descriptionPath = if (forIcon) "description" else "hint"
        val descriptionKey = "game.${gameId.namespace}.${gameId.path}.kit.${kit.id()}.$descriptionPath"

        if (!translations.translator.hasTranslation(translations.getLanguage(player), descriptionKey)) return

        val description = translations.translateText(player, descriptionKey).formatted(ChatFormatting.GREEN)

        val currentLore = stack.getOrDefault(
            DataComponents.LORE,
            ItemLore.EMPTY
        ).styledLines()

        val loreToAdd = IconMaker.wrapText(description, 32)
        val newLore: MutableList<Component> = ArrayList(currentLore)

        if (!currentLore.isEmpty()) {
            // newline between the lore
            newLore.add(Component.nullToEmpty(""))
        }

        newLore.addAll(loreToAdd)

        ItemStackUtil.setLore(stack, newLore)
    }

    fun kitName(kit: Kit): TranslatedText {
        return translations.translateText("game.${gameId.namespace}.${gameId.path}.kit.${kit.id()}")
    }
}
